src/common/typescript.ts

```diff
@@ -0,0 +1,23 @@
+/********************************************************************************
+ * Copyright (C) 2024 EclipseSource and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import { isPlainObject } from 'lodash';
+
+export type RequiredKey<T, K extends keyof T> = T & { [P in K]-?: T[P] };
+
+export function isPlainObjectType(arg: unknown): arg is object {
+    return isPlainObject(arg);
+}
```

Comment:
I've made a lot of comments about utilities, so it may be worth expanding a little on why I keep objecting to them.

The more trivial complaint is that many of the utilities you write are used only once, and in general, that means it's probably too early to extract a utility: we don't know yet in what form it might have general utility.

The more substantial objection has to do with what is idiomatic and correct TypeScript. Because TypeScript freely allows type unions, it's sometimes necessary to check which member of a union type you're dealing with. But it's important that the checks be both expressive and correct. When I say expressive in this case, I mean that it should be clear why the specific checks are made: they should be relevant to distinguishing members of a type union in the context in which it occurs. When I say that checks should be correct, I mean that a type check shouldn't impose unexpected requirements beyond those of a given interface or type definition.

This check is likely to be neither expressive nor correct. It is not correct because it rules out classes:

```
> const lodash = require("lodash")
undefined
> const a = new class {};
undefined
> lodash.isPlainObject(a)
false
```

but in general, it should be possible to implement an interface both as a plain object and as a class. Further, the check isn't likely to be relevant, in that, in the contexts in which it occurs, there is no relevant distinction between an object and some other truthy object-like thing: the type union at issue doesn't have relevant alternatives that this check distinguishes between.

Part of what contributes to this pattern is the use of `unknown`. While it's of course technically correct to design a type guard that could take anything, we write TypeScript and not JavaScript because it allows us to express our expectations about what options will be in play. In this case, what is passed to the `hasMemory` is always a `MemoryState`, and the questions that the `hasMemory` method has to answer are whether the `memory` and `activeReadArguments` fields are the undefined member of their union or not.

But the fact that that the `hasMemory` method is doing two things, only one of which is stated in its name, suggests that something is amiss, as well: it makes sense that memory could be undefined, because fetching that is a very fallible operation, but `activeReadArguments` should never really be undefined, because we know we have a default in the `configuredReadArguments`. That means that in the two places where `activeReadArguments` are set to undefined, you could just set them to the current value of `configuredReadArguments`, since the two are effectively the same. Now `hasMemory` really only needs to check whether `memory` is defined, and that doesn't require a utility function, and without the `hasMemory` function, you also don't need the `RequiredKeys` utility.

TypeScript is very powerful, and there's a lot of scope to extract neat abstract utilities, but every time you're tempted to do so, it's important to ask if that utility really expresses your intent in the clearest and most concise way, or if the perceived need for a utility may indicate that other changes to the code might actually be simpler and clearer.